### PR TITLE
P#51409 Make detail view for reference estates configurable

### PR DESCRIPTION
**Pull request by hungnc89**

### CODE REVIEW

**Review by jayay — COMMENTED** [review]
I don't quite understand - why is "accessControls" stored as an array even though it is a checkbox in the back-end?

File: plugin/Controller/EstateViewDocumentTitleBuilder.php [hunk 1]
```diff
@@ -54,6 +60,11 @@ public function buildDocumentTitle(array $title): array
 		if ($estateId === 0) {
 			return $title;
 		}
+		$accessControl = $this->_pDataDetailViewHandler->getDetailView()->getAccessControls();
+		if (empty($accessControl)) {
+			$title['title'] = 'Error 404 (Not Found)';
```
[the remark this answers is not in the document]
Comment (hungnc89) [reply plugin/Controller/EstateViewDocumentTitleBuilder.php:65]: I use this code in file 'default_detail.php' to handle 404 template but it only works with page body, not title
`if ( empty( $pEstates->getAccessControl() ) ) {
	global $wp_query;
	$wp_query->set_404();
	status_header( 404 );
	get_template_part( 404 );
	exit();
}`

[the remark this answers is not in the document]
Comment (jayay) [reply plugin/Controller/EstateViewDocumentTitleBuilder.php:65]: We've been using

```php
$pWP->handle_404();
include(get_query_template('404'));
die();
```
Where the `$pWP` object is the instance of `WP`.
This has been working fine, but requires the `WP` object.

**Comment by jayay** [comment]
Thank you! 

> I don't quite understand - why is "accessControls" stored as an array even though it is a checkbox in the back-end?

This question is still open.

**Comment by hungnc89** [comment]
> I don't quite understand - why is "accessControls" stored as an array even though it is a checkbox in the back-end?

This was my misunderstanding at the start, but I don't think it has any effect on the functionality I did, so I kept it that way. Do you need me to change it?

**Comment by jayay** [comment]
Thanks, that's a lot cleaner. Could you please also resolve the conflicts?

**Comment by hungnc89** [comment]
@jayay Conflicts have been resolved

**Review by jayay — APPROVED** [review]

File: plugin/EstateList.php [hunk 2]
```diff
@@ -392,6 +395,7 @@ public function estateIterator($modifier = EstateViewFieldModifierTypes::MODIFIE
 
 		$recordModified = $pEstateFieldModifierHandler->processRecord($currentRecord['elements']);
 		$recordRaw = $this->_recordsRaw[$this->_currentEstate['id']]['elements'];
+		$this->setReferenz($recordRaw['referenz']);
```
Comment (jayay) [inline plugin/EstateList.php:398]: `referenz` doesn't always have to be set.

File: plugin/EstateList.php [hunk 3]
```diff
@@ -723,4 +727,12 @@ public function setFormatOutput(bool $formatOutput)
 	/** @return EstateListEnvironment */
 	public function getEnvironment(): EstateListEnvironment
 		{ return $this->_pEnvironment; }
+
+	/** @return string */
+	public function getReferenz(): string
+	{ return $this->_referenz; }
+
+	/** @param string $referenz */
+	public function setReferenz(string $referenz)
```
Comment (jayay) [inline plugin/EstateList.php:743]: Is there a better solution than adding this new property?

Comment (hungnc89) [reply plugin/EstateList.php:743]: @jayay What do you think about adding a new element to the array in the estateIterator function like this?
image

Comment (jayay) [reply plugin/EstateList.php:743]: That looks cleaner, writing `$pArrayContainer['referenz'] = $recordRaw['referenz'] ?? null` would make it a little bit safer, but your proposal could break sites. How about a public interface in `EstateList` to retrieve the `$_recordsRaw`? It should return the raw values inside an `ArrayContainerEscape` as well. If the original value (without escaping, which in turn makes it a string) is needed, it has to be fetched using `$pArrayContainer->getValueRaw()`.

File: css/onoffice_defaultview.css [hunk 4]
```diff
@@ -156,6 +156,12 @@
 	text-decoration: none !important;
 }
 
+.oo-detailslink a:focus {
```
Comment (jayay) [inline css/onoffice_defaultview.css:159]: This could break existing sites, correct?

**Review by jayay — CHANGES_REQUESTED** [review]
Please see my comments. I have some backwards compatibility concerns.

**Comment by hungnc89** [comment]
@jayay I don't know what happened to the git action when I merge master and resolve conflicts, but the unittest still passed.

**Comment by jayay** [comment]
https://github.com/onOfficeGmbH/oo-wp-plugin/pull/106#discussion_r773993261 isn't totally resolved. css/onoffice_defaultview.css gets loaded into existing sites, without a copy in the `templates.dist` folder. That means, changes to that file will appear on existing installations and could break the design.

**Comment by jayay** [comment]
We could introduce a new css file for new design improvements, that resides in the `templates.dist` directory. What do you think?

**Comment by hungnc89** [comment]
@jayay Yes, please introduce me.

**Comment by hungnc89** [comment]
@jayay could you please review this PR?

File: plugin.php [hunk 5]
```diff
@@ -186,8 +188,26 @@
 	$estateId = $pWP->query_vars['estate_id'] ?? '';
 	/** @var EstateIdRequestGuard $pEstateIdGuard */
 	$pEstateIdGuard = $pDI->get(EstateIdRequestGuard::class);
+	/** @var DataDetailViewHandler $pDataDetailViewHandler */
+	$pDataDetailViewHandler = $pDI->get( DataDetailViewHandler::class);
 
 	if ($estateId !== '') {
+		$accessControl = $pDataDetailViewHandler->getDetailView()->getAccessControls();
```
Comment (jayay) [inline plugin.php:195]: This access control check should be an entire new class

File: plugin/DataView/DataDetailViewHandler.php [hunk 6]
```diff
@@ -105,6 +106,7 @@ public function createDetailViewByValues(array $row): DataDetailView
 		$pDataDetailView->setShortCodeForm($row['shortcodeform'] ?? '');
 		$pDataDetailView->setFields($row[DataDetailView::FIELDS] ?? []);
 		$pDataDetailView->setPictureTypes($row[DataDetailView::PICTURES] ?? []);
+		$pDataDetailView->setAccessControls((bool)($row[InputModelOptionFactoryDetailView::INPUT_ACCESS_CONTROL] ?? ''));
```
Comment (jayay) [inline plugin/DataView/DataDetailViewHandler.php:109]: "hasDetailView" would be more understandable here

File: plugin/DataView/DataDetailView.php [hunk 7]
```diff
@@ -158,6 +161,10 @@ public function getFields(): array
 	public function getPictureTypes(): array
 		{ return $this->_pictureTypes; }
 
+	/** @return bool */
+	public function getAccessControls(): bool
```
Comment (jayay) [inline plugin/DataView/DataDetailView.php:165]: "hasDetailView" would be more understandable here

File: plugin/DataView/DataDetailView.php [hunk 8]
```diff
@@ -194,6 +201,10 @@ public function setFields(array $fields)
 	public function setPictureTypes(array $pictureTypes)
 		{ $this->_pictureTypes = $pictureTypes; }
 
+	/** @param bool $accessControl */
+	public function setAccessControls(bool $accessControl)
```
Comment (jayay) [inline plugin/DataView/DataDetailView.php:205]: "setHasDetailView" would be more understandable here

File: plugin/EstateList.php [hunk 9]
```diff
@@ -525,6 +530,18 @@ public function getEstatePictureValues($imageId)
...
 	}
 
+	/**
+	 *
+	 * @return bool
+	 * @throws DependencyException
+	 * @throws NotFoundException
+	 */
+
+	public function getAccessControl(): bool
```
Comment (jayay) [inline plugin/EstateList.php:540]: "hasDetailView" would be more understandable here

**Review by jayay — CHANGES_REQUESTED** [review]
Almost there! There are two things that I annotated

File: plugin/DataView/DataDetailViewCheckAccessControl.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+	private $estateId = null;
+
+
+	/**
+	 * @param int $estateId
+	 */
+
+	public function __construct(int $estateId)
+	{
+		$this->estateId = $estateId;
```
Comment (jayay) [inline plugin/DataView/DataDetailViewCheckAccessControl.php:58]: It'd be nicer if this class was stateless. `int $estateId` would have to be a parameter for `checkAccessControl()`.

**Review by jayay — CHANGES_REQUESTED** [review]

File: plugin.php [hunk 11]
```diff
@@ -186,10 +189,15 @@
 	$estateId = $pWP->query_vars['estate_id'] ?? '';
 	/** @var EstateIdRequestGuard $pEstateIdGuard */
 	$pEstateIdGuard = $pDI->get(EstateIdRequestGuard::class);
+	/** @var DataDetailViewHandler $pDataDetailViewHandler */
+	$pDataDetailViewHandler = $pDI->get( DataDetailViewHandler::class);
 
 	if ($estateId !== '') {
 		$estateId = (int)$estateId;
-		if ($estateId === 0 || !$pEstateIdGuard->isValid($estateId)) {
+		$pDataDetailViewCheckAccessControl = new DataDetailViewCheckAccessControl();
```
Comment (jayay) [inline plugin.php:197]:
```php
/** @var DataDetailViewCheckAccessControl $pDataDetailViewCheckAccessControl */
$pDataDetailViewCheckAccessControl = $pDI->get(DataDetailViewCheckAccessControl::class);
```
woul be better here

File: plugin/DataView/DataDetailViewCheckAccessControl.php [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+
+use onOffice\WPlugin\Factory\EstateListFactory;
+
+/**
+ *
+ * @url http://www.onoffice.de
+ * @copyright 2003-2017, onOffice(R) GmbH
+ *
+ * DO NOT MOVE OR RENAME - NAME AND/OR NAMESPACE MAY BE USED IN SERIALIZED DATA
```
Comment (jayay) [inline plugin/DataView/DataDetailViewCheckAccessControl.php:34]: Is this true?

File: plugin/DataView/DataDetailViewCheckAccessControl.php [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+{
+	/** @var DataDetailViewHandler */
+	private $pDataDetailViewHandler = null;
+
+
+	/**
+	 *
+	 */
+
+	public function __construct()
```
Comment (jayay) [inline plugin/DataView/DataDetailViewCheckAccessControl.php:47]: Dependency `DataDetailViewHandler` could be a constructor argument (follows pattern of dependency injection) and would get injected automatically after small change in plugin.php.

File: plugin/DataView/DataDetailView.php [hunk 14]
```diff
@@ -105,6 +105,9 @@ class DataDetailView
 	/** @var string[] */
 	private $_pictureTypes = [];
 
+	/** @var bool */
+	private $_accessControls = true;
```
Comment (jayay) [inline plugin/DataView/DataDetailView.php:109]: I think some migration code is necessary for this. `$_accessControls` would be `null` in existing views

**Review by jayay — CHANGES_REQUESTED** [review]

File: plugin/DataView/DataDetailViewCheckAccessControl.php (hunk 12 above)
Comment (hungnc89) [reply plugin/DataView/DataDetailViewCheckAccessControl.php:34]: Oh sorry, this isn't true, it's from another class, I copied it :sweat_smile::sweat_smile::sweat_smile:

**Comment by jayay** [comment]
@hungnc89  Would you please also add a test for the class `DataDetailViewCheckAccessControl`?

**Comment by jayay** [comment]
@LongTrong-exe looks good! Could you please pull in the lastest commits from the master branch?

**Comment by LongTrong-exe** [comment]
@jayay here you are.

**Comment by LongTrong-exe** [comment]
@jayay Yes, I've updated the latest commit from master for this branch.

**Review by jayay — APPROVED** [review]

**Comment by michaelarnold-dev** [comment]
@LongTrong-exe Could you please resolve the conflicts? Thank you very much in advance

**Comment by LongTrong-exe** [comment]
@michaelarnold-dev here you are.

**Comment by michaelarnold-dev** [comment]
@LongTrong-exe Thank you! Could you also adjust the code coverage? Thank you in advance!

File: tests/TestClassSDKWrapper.php [hunk 15]
```diff
@@ -134,24 +134,14 @@ public function testAddRequestByApiAction(): APIClientActionGeneric
...
 		$this->assertInstanceOf(DBCache::class, $pDBCaches[0]);
 		$this->assertInstanceOf(onOfficeSDKCache::class, $pDBCaches[0]);
 	}
 
-	/**
-	 *
-	 */
-
-	public function testWriteCache()
-	{
```
Comment (jayay) [inline tests/TestClassSDKWrapper.php]: Has this been broken?

Comment (LongTrong-exe) [reply tests/TestClassSDKWrapper.php]: @jayay Yes, I added it. It worked on my PC, but not on Github.

**Review by jayay — APPROVED** [review]

**Pull request merged**